File: 08-Singleton/src/cpp/main.cpp
#include <iostream>
#include "DatabaseConnection.h"
#include "ConfigManager.h"

int main() {
    std::cout << "=== Singleton Pattern Demo (C++) ===\n\n";

    // 1. DatabaseConnection Singleton
    std::cout << "--- DatabaseConnection Singleton ---\n";
    auto& db1 = DatabaseConnection::get_instance();
    auto& db2 = DatabaseConnection::get_instance();
    std::cout << "Same instance? " << (&db1 == &db2 ? "true" : "false") << "\n";

    db1.connect();
    db2.connect();  // Already connected — same instance!
    std::cout << db1.query("SELECT * FROM users") << "\n";
    db1.disconnect();
    std::cout << "\n";

    // 2. ConfigManager Singleton
    std::cout << "--- ConfigManager Singleton ---\n";
    auto& config1 = ConfigManager::get_instance();
    auto& config2 = ConfigManager::get_instance();
    std::cout << "Same instance? " << (&config1 == &config2 ? "true" : "false") << "\n";
    std::cout << "app.name = " << config1.get("app.name") << "\n";
    config1.set("app.name", "Updated App");
    std::cout << "app.name (via config2) = " << config2.get("app.name") << "\n";
    // Both references see the change — they're the same object

    return 0;
}

File: 08-Singleton/src/go/main.go
package main

import "fmt"

func main() {
	fmt.Println("=== Singleton Pattern Demo (Go) ===")
	fmt.Println()

	// 1. DatabaseConnection Singleton
	fmt.Println("--- DatabaseConnection Singleton ---")
	db1 := GetDatabaseInstance()
	db2 := GetDatabaseInstance()
	fmt.Printf("Same instance? %t\n", db1 == db2)

	db1.Connect()
	db2.Connect() // Already connected — same instance!
	fmt.Println(db1.Query("SELECT * FROM users"))
	db1.Disconnect()
	fmt.Println()

	// 2. ConfigManager Singleton
	fmt.Println("--- ConfigManager Singleton ---")
	config1 := GetConfigInstance()
	config2 := GetConfigInstance()
	fmt.Printf("Same instance? %t\n", config1 == config2)
	fmt.Printf("app.name = %s\n", config1.Get("app.name"))
	config1.Set("app.name", "Updated App")
	fmt.Printf("app.name (via config2) = %s\n", config2.Get("app.name"))
	// Both pointers see the change — they're the same object
}
